### ISSUE TITLE
Shim wishlist

### ISSUE THREAD

**Issue by saethlin** [rust-lang/miri#2057]
I run Miri on the most-downloaded crates, with the flags `-Zmiri-panic-on-unsupported` and `--no-fail-fast` which basically asks Miri to try to run as many tests as it can. Searching all my build logs, I find these are the most-encountered `can't call foreign function: ` errors:
```
15232 counts
(  1)     4329 (28.4%, 28.4%): epoll_create1
(  2)     1128 ( 7.4%, 35.8%): pipe2
(  3)      729 ( 4.8%, 40.6%): socket
(  4)      345 ( 2.3%, 42.9%): TLS_method
(  5)      294 ( 1.9%, 44.8%): __xpg_strerror_r
(  6)      279 ( 1.8%, 46.6%): cuInit
(  7)      257 ( 1.7%, 48.3%): realpath
(  8)      255 ( 1.7%, 50.0%): openat
(  9)      229 ( 1.5%, 51.5%): getaddrinfo
( 10)      205 ( 1.3%, 52.8%): Py_IsInitialized
( 11)      199 ( 1.3%, 54.2%): ruby_init
( 12)      198 ( 1.3%, 55.5%): localtime_r
( 13)      174 ( 1.1%, 56.6%): dlopen
( 14)      164 ( 1.1%, 57.7%): GEOS_init_r
( 15)      161 ( 1.1%, 58.7%): gnu_get_libc_version
( 16)      144 ( 0.9%, 59.7%): eventfd
( 17)      131 ( 0.9%, 60.5%): js_newstate
( 18)      130 ( 0.9%, 61.4%): Rf_initialize_R
( 19)      124 ( 0.8%, 62.2%): secp256k1_context_create
( 20)      111 ( 0.7%, 62.9%): ts_parser_new
( 21)      109 ( 0.7%, 63.6%): mmap
( 22)      106 ( 0.7%, 64.3%): getuid
( 23)      104 ( 0.7%, 65.0%): getpid
( 24)      104 ( 0.7%, 65.7%): lua_newstate
( 25)      103 ( 0.7%, 66.4%): blst_scalar_from_uint64
( 26)      103 ( 0.7%, 67.1%): ion_reader_open_buffer
( 27)      101 ( 0.7%, 67.7%): sigemptyset
( 28)       89 ( 0.6%, 68.3%): gst_init_check
( 29)       88 ( 0.6%, 68.9%): mdb_env_create
( 30)       78 ( 0.5%, 69.4%): OPENSSL_init_ssl
( 31)       68 ( 0.4%, 69.8%): decContextDefault
( 32)       68 ( 0.4%, 70.3%): llvm.x86.addcarry.64
( 33)       65 ( 0.4%, 70.7%): blake3_compress_in_place_sse2
( 34)       58 ( 0.4%, 71.1%): llvm.x86.sse.min.ps
( 35)       57 ( 0.4%, 71.5%): cairo_image_surface_create
( 36)       56 ( 0.4%, 71.8%): GFp_cpuid_setup
( 37)       55 ( 0.4%, 72.2%): psa_crypto_init
( 38)       51 ( 0.3%, 72.5%): SQLAllocHandle
( 39)       50 ( 0.3%, 72.9%): TF_NewGraph
( 40)       50 ( 0.3%, 73.2%): socketpair
( 41)       49 ( 0.3%, 73.5%): getrlimit
( 42)       48 ( 0.3%, 73.8%): rustsecp256k1_v0_4_1_context_preallocated_size
( 43)       48 ( 0.3%, 74.1%): workbook_new
( 44)       44 ( 0.3%, 74.4%): JS_NewRuntime
( 45)       43 ( 0.3%, 74.7%): at_tensor_of_data
( 46)       42 ( 0.3%, 75.0%): cs_open
( 47)       41 ( 0.3%, 75.3%): hoedown_buffer_new
( 48)       41 ( 0.3%, 75.5%): llvm.x86.sse.cmp.ps
( 49)       39 ( 0.3%, 75.8%): curl_global_init
( 50)       33 ( 0.2%, 76.0%): memcpy
```
There are some pretty obvious biases in this data. For example, crates with a lot of tests get their favorite symbols counted more than crates with fewer tests.

`pipe2` and `gnu_get_libc_version` are from tests that try to start a `std::process::Command`. `pipe2` is from `.output()` and `gnu_get_libc_version` is from `spawn()` I think.

I think `__xpg_strerror_r` is from crates that encounter a `std::io::Error` and try to format it. This is pretty common because a lot of crates attempt to access files in their tests that are not packaged (this is also a huge problem for `crater`).

There are also a lot of things in here which I don't think we will ever support via shims. We're not going to have shims for openssl. We're also not going to shim CUDA or JavaScript, and also all those errors are from 3 crates using CUDA and 1 crate doing some JavaScript thing.

But in any case, this is a starting point for prioritizing what shims should be added. Just in case there isn't anyone out there who already has a pet function they want added.

**Comment by RalfJung** [comment]
`__xpg_strerror_r` is something that also comes up when just debugging Miri, I think even some of the tests will debug-print the IO error when the test fails, leading to an annoying double-failure (first the test fails with an assertion, then printing that assertion fails with an "unsupported" error).

**Comment by RalfJung** [comment]
> I think __xpg_strerror_r is from crates that encounter a std::io::Error and try to format it. This is pretty common because a lot of crates attempt to access files in their tests that are not packaged (this is also a huge problem for crater).

Wish granted. ;) https://github.com/rust-lang/miri/pull/2067

**Comment by mejrs** [comment]
`( 10)      205 ( 1.3%, 52.8%): Py_IsInitialized`

This the first CPython api call that Rust programs embedding the python interpreter will do. I would really like to test `pyo3` and these programs with Miri.

I know Python's C api is too big to shim, but I would love to see some solution for this.

**Comment by CraftSpider** [comment]
I have a similar wishlist for testing a PHP extension library. Currently it fails at `pipe2`, but after that it would start failing at whatever the first call to PHP is. It would be nice if there was some way to use miri to test the Rust parts of my project, and then just actually call out to the relevant PHP stuff as normal.

**Comment by jjl** [comment]
mmap() is the one i'm currently missing. for my purposes it could just use a boxed slice. would be a little painful on memory, but enough to get miri running.

**Comment by bjorn3** [comment]
Read-only mmap where you don't write to the backing file shouldn't be too hard to implement. If writes through either the mapping or through the backing file are allowed, it becomes a lot harder to implement unix compatible semantics.

**Comment by jjl** [comment]
Ah, I see what you mean. Fortunately, I'm only using anonymous mappings.

**Comment by LegNeato** [comment]
Thanks for the list!  I picked the easiest one I could see near the top ([realpath](https://github.com/rust-lang/miri/pull/2294)):

Is there any way we can get a couple of example crates for each missing shim in the list so implementers can check the before and after with real code?

**Comment by saethlin** [comment]
I've manually grabbed a list for you for `realpath`: [realpath-crates.txt](https://github.com/rust-lang/miri/files/9022085/realpath-crates.txt)

I have a very simple website which is just Cloudfront + S3 at https://miri.saethlin.dev (most people care about https://miri.saethlin.dev/ub). I'm not opposed to making simple additions, like a way to download all the raw logs in a `.tar.gz`. You can comment here or open a new issue: https://github.com/saethlin/miri-tools/issues/8

**Comment by saethlin** [comment]
Since @eduardosm has been doing some amazing work on implementing LLVM x86 intrinsics, here's my list of those that are encountered currently in test suites:
```
(  1)     6867 (74.6%, 74.6%): llvm.x86.subborrow.64
(  2)      539 ( 5.9%, 80.5%): llvm.x86.sse2.psad.bw
(  3)      361 ( 3.9%, 84.4%): llvm.x86.sse.cmp.ps
(  4)      177 ( 1.9%, 86.3%): llvm.x86.sse2.cmp.pd
(  5)      137 ( 1.5%, 87.8%): llvm.x86.sse.add.ss
(  6)       93 ( 1.0%, 88.8%): llvm.x86.aesni.aesenc
(  7)       87 ( 0.9%, 89.8%): llvm.x86.sse.sqrt.ss
(  8)       86 ( 0.9%, 90.7%): llvm.x86.rdtsc
(  9)       74 ( 0.8%, 91.5%): llvm.prefetch
( 10)       64 ( 0.7%, 92.2%): llvm.x86.bmi.pdep.64
( 11)       60 ( 0.7%, 92.9%): llvm.x86.sse2.pmulu.dq
( 12)       57 ( 0.6%, 93.5%): llvm.x86.sse2.pslli.d
( 13)       54 ( 0.6%, 94.1%): llvm.x86.sse.sqrt.ps
( 14)       40 ( 0.4%, 94.5%): llvm.x86.sse.movmsk.ps
( 15)       36 ( 0.4%, 94.9%): llvm.x86.sse.min.ps
( 16)       33 ( 0.4%, 95.3%): llvm.x86.sse.max.ps
( 17)       26 ( 0.3%, 95.6%): llvm.x86.sse2.cvttps2dq
( 18)       24 ( 0.3%, 95.8%): llvm.x86.sse.cmp.ss
( 19)       24 ( 0.3%, 96.1%): llvm.x86.sse2.cmp.sd
( 20)       22 ( 0.2%, 96.3%): llvm.x86.sse2.psll.q
( 21)       19 ( 0.2%, 96.5%): llvm.x86.sse2.psll.d
( 22)       18 ( 0.2%, 96.7%): llvm.x86.sse2.cvtps2dq
( 23)       17 ( 0.2%, 96.9%): llvm.x86.pclmulqdq
( 24)       17 ( 0.2%, 97.1%): llvm.x86.sse2.psrl.d
( 25)       16 ( 0.2%, 97.3%): llvm.x86.sse2.psrl.q
( 26)       15 ( 0.2%, 97.4%): llvm.x86.sse2.storeu.dq
( 27)       10 ( 0.1%, 97.5%): llvm.x86.sse.stmxcsr
( 28)       10 ( 0.1%, 97.6%): llvm.x86.sse2.pmulh.w
( 29)       10 ( 0.1%, 97.8%): llvm.x86.sse2.psll.w
( 30)       10 ( 0.1%, 97.9%): llvm.x86.sse2.pslli.w
( 31)        9 ( 0.1%, 98.0%): llvm.x86.sse2.movmsk.pd
( 32)        8 ( 0.1%, 98.0%): llvm.x86.sse2.psra.d
( 33)        6 ( 0.1%, 98.1%): llvm.x86.sse.rcp.ps
( 34)        6 ( 0.1%, 98.2%): llvm.x86.sse.rsqrt.ps
( 35)        6 ( 0.1%, 98.2%): llvm.x86.sse2.cvtdq2ps
( 36)        6 ( 0.1%, 98.3%): llvm.x86.sse2.max.pd
( 37)        6 ( 0.1%, 98.4%): llvm.x86.sse2.min.pd
( 38)        6 ( 0.1%, 98.4%): llvm.x86.sse2.psra.w
( 39)        6 ( 0.1%, 98.5%): llvm.x86.sse2.psrli.d
( 40)        5 ( 0.1%, 98.6%): llvm.x86.sse2.packsswb.128
( 41)        5 ( 0.1%, 98.6%): llvm.x86.sse2.packuswb.128
( 42)        5 ( 0.1%, 98.7%): llvm.x86.ssse3.pshuf.b.128
( 43)        4 ( 0.0%, 98.7%): llvm.x86.sse.sub.ss
( 44)        4 ( 0.0%, 98.7%): llvm.x86.sse2.comige.sd
( 45)        4 ( 0.0%, 98.8%): llvm.x86.sse2.psrl.w
( 46)        4 ( 0.0%, 98.8%): llvm.x86.sse3.ldu.dq
( 47)        3 ( 0.0%, 98.9%): llvm.x86.sse2.packssdw.128
( 48)        3 ( 0.0%, 98.9%): llvm.x86.sse2.psrai.d
( 49)        3 ( 0.0%, 98.9%): llvm.x86.sse2.psrai.w
( 50)        3 ( 0.0%, 99.0%): llvm.x86.sse2.psrli.q
( 51)        2 ( 0.0%, 99.0%): llvm.ctpop.v16i8
( 52)        2 ( 0.0%, 99.0%): llvm.x86.avx2.pslli.d
( 53)        2 ( 0.0%, 99.0%): llvm.x86.sse.comieq.ss
( 54)        2 ( 0.0%, 99.1%): llvm.x86.sse.comige.ss
( 55)        2 ( 0.0%, 99.1%): llvm.x86.sse.comigt.ss
( 56)        2 ( 0.0%, 99.1%): llvm.x86.sse.comile.ss
( 57)        2 ( 0.0%, 99.1%): llvm.x86.sse.comilt.ss
( 58)        2 ( 0.0%, 99.1%): llvm.x86.sse.comineq.ss
( 59)        2 ( 0.0%, 99.2%): llvm.x86.sse.cvtsi2ss
( 60)        2 ( 0.0%, 99.2%): llvm.x86.sse.cvtss2si
( 61)        2 ( 0.0%, 99.2%): llvm.x86.sse.div.ss
( 62)        2 ( 0.0%, 99.2%): llvm.x86.sse.max.ss
( 63)        2 ( 0.0%, 99.2%): llvm.x86.sse.min.ss
( 64)        2 ( 0.0%, 99.3%): llvm.x86.sse.mul.ss
( 65)        2 ( 0.0%, 99.3%): llvm.x86.sse.rcp.ss
( 66)        2 ( 0.0%, 99.3%): llvm.x86.sse.rsqrt.ss
( 67)        2 ( 0.0%, 99.3%): llvm.x86.sse2.comieq.sd
( 68)        2 ( 0.0%, 99.4%): llvm.x86.sse2.comile.sd
( 69)        2 ( 0.0%, 99.4%): llvm.x86.sse2.comilt.sd
( 70)        2 ( 0.0%, 99.4%): llvm.x86.sse2.comineq.sd
( 71)        2 ( 0.0%, 99.4%): llvm.x86.sse2.cvtpd2dq
( 72)        2 ( 0.0%, 99.4%): llvm.x86.sse2.cvtpd2ps
( 73)        2 ( 0.0%, 99.5%): llvm.x86.sse2.cvtps2pd
( 74)        2 ( 0.0%, 99.5%): llvm.x86.sse2.cvtsd2si
( 75)        2 ( 0.0%, 99.5%): llvm.x86.sse2.cvtsd2si64
( 76)        2 ( 0.0%, 99.5%): llvm.x86.sse2.cvtsd2ss
( 77)        2 ( 0.0%, 99.6%): llvm.x86.sse2.cvtss2sd
( 78)        2 ( 0.0%, 99.6%): llvm.x86.sse2.cvttpd2dq
( 79)        2 ( 0.0%, 99.6%): llvm.x86.sse2.cvttsd2si
( 80)        2 ( 0.0%, 99.6%): llvm.x86.sse2.cvttsd2si64
( 81)        2 ( 0.0%, 99.6%): llvm.x86.sse2.max.sd
( 82)        2 ( 0.0%, 99.7%): llvm.x86.sse2.min.sd
( 83)        2 ( 0.0%, 99.7%): llvm.x86.sse2.pavg.b
( 84)        2 ( 0.0%, 99.7%): llvm.x86.sse2.pavg.w
( 85)        2 ( 0.0%, 99.7%): llvm.x86.sse2.pmadd.wd
( 86)        2 ( 0.0%, 99.7%): llvm.x86.sse2.pmulhu.w
( 87)        2 ( 0.0%, 99.8%): llvm.x86.sse2.pslli.q
( 88)        2 ( 0.0%, 99.8%): llvm.x86.sse2.psrli.w
( 89)        2 ( 0.0%, 99.8%): llvm.x86.sse2.sqrt.sd
( 90)        2 ( 0.0%, 99.8%): llvm.x86.sse2.storeu.pd
( 91)        2 ( 0.0%, 99.9%): llvm.x86.sse3.hadd.ps
( 92)        2 ( 0.0%, 99.9%): llvm.x86.sse41.dpps
( 93)        2 ( 0.0%, 99.9%): llvm.x86.sse41.ptestz
( 94)        1 ( 0.0%, 99.9%): llvm.ctpop.v2i64
( 95)        1 ( 0.0%, 99.9%): llvm.ctpop.v4i32
( 96)        1 ( 0.0%, 99.9%): llvm.ctpop.v8i16
( 97)        1 ( 0.0%, 99.9%): llvm.x86.avx2.permd
( 98)        1 ( 0.0%,100.0%): llvm.x86.avx2.pslli.q
( 99)        1 ( 0.0%,100.0%): llvm.x86.bmi.bzhi.64
(100)        1 ( 0.0%,100.0%): llvm.x86.fma.vfnmadd.pd.256
(101)        1 ( 0.0%,100.0%): llvm.x86.rdrand.64
(102)        1 ( 0.0%,100.0%): llvm.x86.rdseed.64
```

**Comment by saethlin** [comment]
Here's an update:
```
(  1)      214 (40.9%, 40.9%): llvm.x86.sse41.pblendw
(  2)      128 (24.5%, 65.4%): llvm.x86.rdtsc
(  3)       37 ( 7.1%, 72.5%): llvm.x86.sse41.blendvps
(  4)       22 ( 4.2%, 76.7%): llvm.x86.sse41.ptestz
(  5)       19 ( 3.6%, 80.3%): llvm.x86.sse42.crc32.32.8
(  6)       17 ( 3.3%, 83.6%): llvm.x86.sse42.pcmpestri128
(  7)       12 ( 2.3%, 85.9%): llvm.x86.sse.stmxcsr
(  8)       10 ( 1.9%, 87.8%): llvm.x86.sse41.round.ps
(  9)        7 ( 1.3%, 89.1%): llvm.x86.sse41.packusdw
( 10)        6 ( 1.1%, 90.2%): llvm.x86.sse41.round.sd
( 11)        6 ( 1.1%, 91.4%): llvm.x86.sse41.round.ss
( 12)        5 ( 1.0%, 92.4%): llvm.x86.sse41.dpps
( 13)        4 ( 0.8%, 93.1%): llvm.x86.sse41.insertps
( 14)        3 ( 0.6%, 93.7%): llvm.x86.sse41.mpsadbw
( 15)        3 ( 0.6%, 94.3%): llvm.x86.sse41.phminposuw
( 16)        3 ( 0.6%, 94.8%): llvm.x86.sse41.pmuldq
( 17)        3 ( 0.6%, 95.4%): llvm.x86.sse41.ptestc
( 18)        3 ( 0.6%, 96.0%): llvm.x86.sse41.ptestnzc
( 19)        3 ( 0.6%, 96.6%): llvm.x86.sse41.round.pd
( 20)        3 ( 0.6%, 97.1%): llvm.x86.sse42.crc32.32.16
( 21)        3 ( 0.6%, 97.7%): llvm.x86.sse42.crc32.32.32
( 22)        3 ( 0.6%, 98.3%): llvm.x86.sse42.crc32.64.64
( 23)        3 ( 0.6%, 98.9%): llvm.x86.sse42.pcmpestrm128
( 24)        3 ( 0.6%, 99.4%): llvm.x86.sse42.pcmpistri128
( 25)        2 ( 0.4%, 99.8%): llvm.experimental.vector.reduce.add.v4i128
( 26)        1 ( 0.2%,100.0%): llvm.x86.sse41.blendvpd
```
I do not know what happened to the `aesni.aesenc` calls. My best guess is that I changed to `-Ctarget-cpu=x86_64-v2` and now they're hidden behind some SSE4 functions.

**Comment by def-** [comment]
Thanks for posting this. We use miri in [Materialize](https://github.com/MaterializeInc/materialize) and seem to be having a different set of unsupported operations, basically our wishlist:
```
$ git grep "cfg_attr(miri"|sed -e "s#.*\/\/ *##" | sed -e "s/error: *//" | grep -v "slow"|sort|uniq -c|sort -n -r
     81 unsupported operation: can't call foreign function `TLS_client_method` on OS `linux`
     58 unsupported operation: returning ready events from epoll_wait is not yet implemented
     20 unsupported operation: can't call foreign function `decNumberFromInt32` on OS `linux`
     15 unsupported operation: can't call foreign function `decContextDefault` on OS `linux`
     12 unsupported operation: can't call foreign function `rust_psm_stack_pointer` on OS `linux`
     12 unsupported operation: cannot write to event
     11 unsupported operation: can't call foreign function `OPENSSL_init_ssl` on OS `linux`
      7 unsupported operation: integer-to-pointer casts and `ptr::from_exposed_addr` are not supported with `-Zmiri-strict-provenance`
      6 unsupported operation: inline assembly is not supported
      6 unsupported operation: can't call foreign function `TLS_method` on OS `linux`
      4 unsupported operation: can't call foreign function `socket` on OS `linux`
      4 unsupported operation: can't call foreign function `pipe2` on OS `linux`
      3 unsupported operation: can't call foreign function `rocksdb_create_default_env` on OS `linux`
      3 unsupported operation: can't call foreign function `epoll_wait` on OS `linux`
      2 unsupported operation: can't call foreign function `deflateInit2_` on OS `linux`
      1 unsupported operation: non-default mode 0o600 is not supported
      1 unsupported operation: integer-to-pointer casts and `ptr::from_exposed_addr`
      1 unsupported operation: can't call foreign function `pidfile_open` on OS `linux`
```

**Comment by newpavlov** [comment]
My list of desired shims: `O_NOFOLLOW` flag support for `open`, `flock` with all 3 flags, `pwrite64`, and `pread64`. It's not a full list, just functions on which our tests are stuck (we can `cfg` our way out for the former two, but not for the latter two).

**Comment by RalfJung** [comment]
@saethlin it might be time to update the list for FFI shims as well -- the two first items in the list at the top of this issue are now implemented in Miri. :D

`socket` is still missing so that will probably be number 1 now. (Or all those `pipe2` where actually from `process::Command` and we'll see a bunch of `fork`/`posix_spawn`-like things.)

**Comment by saethlin** [comment]
I've updated the issue description with all the crates I've run this month, which is the ~32,000 crates with the most downloads in the last 90 days.

I'm quite surprised that we don't see a bunch of `process::Command` internals as you suggested. Maybe I'll check in on that later?

**Comment by wucke13** [comment]
For us over at https://github.com/rosenpass, the most predominant issues are `mprotect` and to some extent `mmap`. Our use case are custom allocators for use in cryptography, so a place where manual memory management and a desire for high assurance converge. Working around it using conditional compilation is certainly possible, but introduces a non-neglectable delta between the miri-vetted and the release code.

**Comment by RalfJung** [comment]
Regarding mmap, we do support some very simple uses of it. https://github.com/rust-lang/miri/issues/3605 tracks supporting more. However there are also some common ways of using mmap that are technically unsound in Rust (and in every LLVM-based compiler) and need more foundational work first. They will often work in practice, but in Miri we don't want to accept code that is technically unsound even if it works in practice.

Regarding mprotect, that too has some gnarly soundness question which I don't think have even been discussed yet (Miri itself uses it in FFI mode and the soundness reasoning is non-trivial). I opened https://github.com/rust-lang/unsafe-code-guidelines/issues/580 to track that.

**Comment by Bougerous** [comment]
Another commonly hit one on macOS: `CFLocaleCopyPreferredLanguages` — triggered by `sys-locale` crate (used by `cosmic-text`, a popular text rendering library). Any project using `cosmic-text` for font/text handling can't run under Miri on macOS due to this.

Understand this is macOS-specific CoreFoundation FFI so may not fit the shim criteria, but noting the impact.

**Comment by RalfJung** [comment]
I assume that any text rendering library will need tons of APIs for actually doing graphics, and `CFLocaleCopyPreferredLanguages` is just the first function you are seeing. Graphics APIs are definitely out of scope, so I would be surprised if implementing `CFLocaleCopyPreferredLanguages` would actually unlock anything -- but please correct me if I am wrong.

**Comment by CraftSpider** [comment]
`sys-locale`, specifically, would only need this set of CoreFoundation APIs:
```
#[link(name = "CoreFoundation", kind = "framework")]
extern "C" {
    fn CFArrayGetCount(theArray: CFArrayRef) -> CFIndex;
    fn CFArrayGetValueAtIndex(theArray: CFArrayRef, idx: CFIndex) -> *const c_void;

    fn CFStringGetLength(theString: CFStringRef) -> CFIndex;
    fn CFStringGetBytes(
        theString: CFStringRef,
        range: CFRange,
        encoding: CFStringEncoding,
        lossByte: u8,
        isExternalRepresentation: Boolean,
        buffer: *mut u8,
        maxBufLen: CFIndex,
        usedBufLen: *mut CFIndex,
    ) -> CFIndex;

    fn CFRelease(cf: CFTypeRef);

    fn CFLocaleCopyPreferredLanguages() -> CFArrayRef;
}
```

`cosmic-text` is largely pure-Rust, relying on HarfRust for actual shaping, and the `fontations` project for font file handling, so _may_ actually work if miri were to support some of the basic CoreFoundation APIs. I did a quick dive through dependencies and didn't see any immediate red flags, aside from an optional transitive dep on a full terminal emulator, but that's optional and I suspect unrelated to the core text functionality. So overall, I think it's worth considering whether miri wants to support enough CoreFoundation for meaningful programs.

Below the cut is some more in-depth thoughts on CoreFoundation and what's involved in supporting it
<details>
  <summary>CoreFoundation thoughts</summary>

CoreFoundation is a macos framework, which is defined in more depth [here](https://developer.apple.com/library/archive/documentation/General/Conceptual/DevPedia-CocoaCore/Framework.html). Basically, it's a shared library plus some other stuff. They're generally defined in the documentation in terms of either Objective C or Swift, but are in the end still linkable libraries for our purposes.

The frameworks are largely designed around a semi-OOP style, where most functions pass around pointers to structs which all, at least conceptually, subclass `CFType`. So every type (`CFString`, `CFArray`, etc) has a `CFTypeID`, using which, it can be safely converted to and from the universal supertype. This means arrays and are dictionaries are generally homogenous, not caring about what they store. The exception is that for most such storage types, you can optionally pass in a custom VTable on creation which handles the ownership lifecycle of contained objects. Meaning from the Rust side, you could create a `CFArray` which managed some `Box<Type>`, and when that box was removed would call some `extern "C"` shim that drops the box.

All this means that Miri has an important consideration: If it were to implement a subset, how much? It might be easier to implement a drop-in replacement in pure Rust than in the language of Miri, since that's basically what it is. Just a standard library for MacOS code[^1]. Point is, just the 6 functions above could practically be a single PR, but if you want to support any meaningful chunk of it, things get complicated. And that's not counting other frameworks - there's a whole set of them. CoreText, CoreGraphics, CoreImage, etc. Obviously most of those would be out of scope, but there's a non-zero chance people ask for them if we support `CoreFoundation`.

[^1]: Having had this idea while writing this entire thing, I wonder how hard that would be to do - allow pointing to Rust libraries that expose various functions as an alternative to shims, and have miri link those into the execution only when running under Miri.

</details>

**Comment by RalfJung** [comment]
> I wonder how hard that would be to do - allow pointing to Rust libraries that expose various functions as an alternative to shims, and have miri link those into the execution only when running under Miri. 

It's already possible to implement those shims in Rust. Anywhere in your code you can do
```rust
#[cfg(miri)] // for the "use only when running under Miri" part
#[link_name = "malloc"]
fn my_malloc(...) { ... }
```
and Miri will invoke that function whenever `malloc` is called. Obviously that's a bad idea for `malloc`, but you can put the link names of these CoreFoundation symbols there and play around with that.

That makes me wonder if we should maybe do this for all these vendor intrinsics, many of them can be implemented entirely in Rust just fine...

**Comment by CraftSpider** [comment]
To the original request - quickly stubbing out the CoreFoundation APIs, we hit 'Miri does not support file-backed memory mappings'. So this is meaningfully blocked on `mmap` of files, see #3605.

**Comment by Bougerous** [comment]
@CraftSpider Thanks for the confirmation , I'll not pursue this further
